File: src/main/java/net/maple/handlers/user/UserTransferFieldRequestHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import client.messages.broadcast.types.AlertMessage;
import field.object.portal.FieldPortal;
import net.maple.handlers.PacketHandler;
import net.maple.packets.CharacterPackets;
import util.packet.PacketReader;

public class UserTransferFieldRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();

        boolean cashShop = !reader.readBool();
        if (cashShop) {
            if (!chr.isInCashShop()) {
                c.close(this, "Not in CashShop");
                return;
            }
            chr.setInCashShop(false);
            c.migrate();
        } else {
            int id = reader.readInteger();

            if (id != -1) {
                if (c.isAdmin()) {
                    chr.changeField(id);
                } else {
                    c.close(this, "Using /m without admin acc");
                }
            } else {
                String portalName = reader.readMapleString();
                FieldPortal portal = chr.getField().getPortalByName(portalName);

                if (portal == null) {
                    chr.enableActions();
                    chr.write(CharacterPackets.message(new AlertMessage("There is a problem with the portal!\r\nName: " + portalName)));
                    return;
                }
                portal.enter(chr);
            }
        }
    }
}

File: src/main/java/net/maple/handlers/user/UserTransferFieldRequestHandler.kt
package net.maple.handlers.user

import client.Client
import client.messages.broadcast.types.AlertMessage
import constants.FieldConstants
import constants.FieldConstants.isSafeMap
import net.maple.handlers.PacketHandler
import net.maple.packets.CharacterPackets.message
import util.HexTool
import util.logging.LogType
import util.logging.Logger
import util.packet.PacketReader
import java.awt.Point

class UserTransferFieldRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        println(HexTool.toHex(reader.data))
        val chr = c.character
        val cashShop = reader.available() == 0

        if (cashShop) {
            if (!chr.isInCashShop) {
                c.close(this, "Not in CashShop")
                return
            }

            chr.isInCashShop = false
            c.migrate()
        } else {
            val fieldKey = reader.readByte()
            if (fieldKey != chr.fieldKey) {
                return run {
                    Logger.log(LogType.FIELD_KEY, "FieldKey mismatch", this, c)
                }
            }

            val fieldId = reader.readInteger()
            val portalName = reader.readMapleString()
            if (fieldId != -1) {
                val point = if (portalName.isNotEmpty()) {
                    reader.readPoint()
                } else {
                    Point(0, 0)
                }

                val townPortal = reader.readByte()
                val premium = reader.readBool()

                if (c.isAdmin && chr.isAlive()) {
                    val chase = reader.readBool()
                    if (chase) {
                        chr.chasing = true
                        chr.position = Point(reader.readInteger(), reader.readInteger())
                    }
                    chr.changeField(fieldId)
                } else if (!chr.isAlive()) {
                    chr.health = 50
                    if (chr.hardcore && !chr.field.template.isSafeMap() && !chr.safeDeath) {
                        chr.changeField(FieldConstants.HARDCORE_DEATHMAP)
                    } else {
                        chr.changeField(chr.field.template.returnMap)
                        chr.safeDeath = false
                    }
                } else {
                    Logger.log(LogType.HACK, "Using /m or /c without admin acc", this, c)
                    c.close()
                }
            } else {
                val portal = chr.field.getPortalByName(portalName) ?: return run {
                    chr.enableActions()
                    chr.message(AlertMessage("There is a problem with the portal!\r\nName: $portalName"))
                }

                portal.enter(chr)
            }
        }
    }
}